engine/src/main/java/org/terasology/engine/subsystem/openvr/OpenVRControllers.java

```diff
@@ -0,0 +1,171 @@
+ * limitations under the License.
+ */
+package org.terasology.engine.subsystem.openvr;
+
+import jopenvr.VRControllerState_t;
+import org.joml.Matrix4f;
+import org.terasology.input.ButtonState;
+import org.terasology.input.ControllerDevice;
+import org.terasology.input.InputType;
+import org.terasology.input.device.ControllerAction;
+import org.terasology.rendering.openvrprovider.ControllerListener;
+import org.terasology.rendering.openvrprovider.OpenVRUtil;
+import org.terasology.input.ControllerId;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+
+/**
+ * This class acts as an interface between OpenVR controllers and Terasology controllers. By implementing
+ * ControllerListener.buttonStateChanged(), OpenVR controller events make it into this class and are stored. Then,
+ * by implementing getInputQueue(), an overridden method from ControllerDevice, the Terasology controller system
+ * frequently retrieves and handles these events.
+ */
+class OpenVRControllers implements ControllerDevice, ControllerListener {
+
+    private Queue<ControllerAction> queuedActions = new ArrayDeque<>();
+    private VRControllerState_t cachedStateBefore;
+    private VRControllerState_t cachedStateAfter;
+
+    /**
+     * Get the controller names provided by this ControllerDevice.
+     * @return the list of controllers names.
+     */
+    @Override
+    public List<String> getControllers() {
+        List<String> ids = new ArrayList<>();
+        ids.add("OpenVR");
+        return ids;
+    }
+
+    /**
+     * Get all queued actions registered since this method was last called.
+     * @return a queue of actions.
+     */
+    @Override
+    public Queue<ControllerAction> getInputQueue() {
+        Queue<ControllerAction> result = new ArrayDeque<>(queuedActions);
+        queuedActions.clear();
+        return result;
+    }
+
+    private boolean buttonUp(long buttonIndex) {
+        return OpenVRUtil.switchedUp(buttonIndex, cachedStateBefore.ulButtonPressed, cachedStateAfter.ulButtonPressed);
+    }
+
+    private boolean buttonDown(long buttonIndex) {
+        return OpenVRUtil.switchedDown(buttonIndex, cachedStateBefore.ulButtonPressed, cachedStateAfter.ulButtonPressed);
+    }
+
+    private void addAxisAction(int controllerButton, ButtonState buttonState, float axisValue) {
+        queuedActions.add(new ControllerAction(InputType.CONTROLLER_AXIS.getInput(controllerButton),
+                    "OpenVR", buttonState, axisValue));
+    }
+
+    private void addButtonAction(int controllerButton, ButtonState buttonState) {
+        queuedActions.add(new ControllerAction(InputType.CONTROLLER_BUTTON.getInput(controllerButton),
+                "OpenVR", buttonState, 1.0f));
+    }
+
+    private void handleController0() {
+        if (buttonUp(ControllerListener.BUTTON_TRIGGER)) {
+            addButtonAction(ControllerId.ZERO, ButtonState.UP);
```

Comment:
1. Iterating through xor of bit mask:
By doing what you suggest I think we could knock down a few lines of code, but I personally think it would end up being a little less readable. That sort of if statement block, while a bit ugly, is an idiom I've seen many times. Since the efficiency difference is minimal, I personally think that sacrificing a little bit of efficiency for readability would be better.

2. ControllerId.ZERO->ButtonId.ZERO: this goes once again into other folks' code. If we're fine with me editing those defines throughout the whole codebase, I'm on board for refactoring it.

3. BUTTON_TRIGGER->TRIGGER_BUTTON: that I can do. It'll be in my next commit.